eng/scripts/live-test-resource-cleanup.ps1

```diff
@@ -3,93 +3,373 @@
+  }
+  if (!$users) {
+    Write-Error "Failed to retrieve github -> microsoft alias mappings from opensource api."
+    exit 1
+  }
+  foreach ($user in $users) {
+    if ($user.aad.alias) {
+      $OwnerAliasCache[$user.aad.alias] = $true
+    }
+  }
+  Write-Host "Found $($OwnerAliasCache.Count) github+microsoft user records."
+  if ($GithubAliasCachePath -and !(Test-Path $GithubAliasCachePath)) {
+      Write-Host "Caching github -> microsoft alias mappings to '$GithubAliasCachePath'"
+      $users | ConvertTo-Json -Depth 100 | Out-File $GithubAliasCachePath -WhatIf:$false
+  }
+}
 
-$now = [DateTime]::UtcNow
+function GetTag([object]$ResourceGroup, [string]$Key) {
+  if (!$ResourceGroup.Tags) {
+    return $null
+  }
 
-$noDeleteAfter = $allGroups.Where({ $_.Tags.Keys -notcontains "DeleteAfter" })
-Write-Host "Subscription contains $($noDeleteAfter.Count) resource groups with no DeleteAfter tags"
-$noDeleteAfter | ForEach-Object { Write-Verbose $_.ResourceGroupName }
+  foreach ($tagKey in $ResourceGroup.Tags.Keys) {
+    # Compare case-insensitive
+    if ($tagKey -ieq $Key) {
+      return $ResourceGroup.Tags[$tagKey]
+    }
+  }
 
-$hasDeleteAfter = $allGroups.Where({ $_.Tags.Keys -contains "DeleteAfter" })
-Write-Host "Count $($hasDeleteAfter.Count)"
-$toDelete = $hasDeleteAfter.Where({ $deleteDate = ($_.Tags.DeleteAfter -as [DateTime]); (!$deleteDate -or $now -gt $deleteDate) })
-Write-Host "Groups to delete: $($toDelete.Count)"
+  return $null
+}
 
-# Get purgeable resources already in a deleted state.
-$purgeableResources = @(Get-PurgeableResources)
+function HasValidOwnerTag([object]$ResourceGroup) {
+  $ownerTag = GetTag $ResourceGroup "Owners"
+  if (!$ownerTag) {
+    return $false
+  }
+  $owners = $ownerTag -split "[;, ]"
+  $hasValidOwner = $false
+  $invalidOwners = @()
+  foreach ($owner in $owners) {
+    if (IsValidAlias -Alias $owner) {
+      $hasValidOwner = $true
+    } else {
+      $invalidOwners += $owner
+    }
+  }
+  if ($invalidOwners) {
+    Write-Warning " Resource group '$($ResourceGroup.ResourceGroupName)' has invalid owner tags: $($invalidOwners -join ',')"
+  }
+  if ($hasValidOwner) {
+    Write-Host " Skipping tagged resource group '$($ResourceGroup.ResourceGroupName)' with owners '$owners'"
+    return $true
+  }
+  return $false
+}
 
-foreach ($rg in $toDelete)
-{
-  if ($Force -or $PSCmdlet.ShouldProcess("$($rg.ResourceGroupName) (UTC: $($rg.Tags.DeleteAfter))", "Delete Group")) {
-    # Add purgeable resources that will be deleted with the resource group to the collection.
-    $purgeableResourcesFromRG = Get-PurgeableGroupResources $rg.ResourceGroupName
+function HasValidAliasInName([object]$ResourceGroup) {
+    # check compliance (formatting first, then validate alias) and skip if compliant
+    if ($ResourceGroup.ResourceGroupName `
+      -match '^(rg-)?(?<alias>(t-|a-|v-)?[a-z,A-Z]+)([-_]{1}.*)?$' `
```

Comment:
`[-_]{1}` Why is `{1}` interesting in this case as compared to just `[-_]`?

Reply:
That's a great question. It was in the original regex I copied over, I can try without it. It does seem redundant.